### PR TITLE
`merkledb` -- preallocate `bytes.Buffer` in codec

### PR DESCRIPTION
**Pull request by danlaine**
## Why this should be merged

May prevent doing some memory allocations

## How this works

Self-explanatory

## How this was tested

Existing UT

### CODE REVIEW

File: x/merkledb/codec.go [hunk 1]
```diff
@@ -81,7 +84,7 @@ type codecImpl struct {
 }
 
 func (c *codecImpl) encodeDBNode(n *dbNode) []byte {
-	buf := &bytes.Buffer{}
+	buf := bytes.NewBuffer(make([]byte, 0, defaultNodeAllocSize))
```
Comment (joshua-kim) [inline x/merkledb/codec.go:87]: I actually wonder if it's possible to get a tighter estimate on this preallocation. the `dbNode` has two fields which is a `maybe.Maybe[[]byte]` and a `map[byte]child` where `child` is type of `path (string)` and `ids.ID (32 bytes)`.

Is it possible for us to do add up the lengths of these at runtime and that would give us a better allocation estimate..? I'm not super familiar with this code so I ignore me if I'm saying something that sounds strange X)

File: x/merkledb/codec.go [hunk 2]
```diff
@@ -97,7 +100,7 @@ func (c *codecImpl) encodeDBNode(n *dbNode) []byte {
 }
 
 func (c *codecImpl) encodeHashValues(hv *hashValues) []byte {
-	buf := &bytes.Buffer{}
+	buf := bytes.NewBuffer(make([]byte, 0, defaultHashValuesAllocSize))
```
Comment (joshua-kim) [inline x/merkledb/codec.go:103]: Similar comment as above

**Comment by joshua-kim** [comment]
Code looks correct to me but I'm just curious to see if we can get tighter bounds on the pre-allocation to see if we can avoid re-allocations later.

**Comment by danlaine** [comment]
@joshua-kim Added estimates based on the number of children

**Comment by danlaine** [comment]
For what it's worth I expect this to be a pretty modest speedup. If we find that we're doing a ton of memory allocations here we can always bump the size estimates.

**Review by dhrubabasu — APPROVED** [review]

**Pull request merged**